File: demo-formatter/go/cucumber_demo_formatter.go
/*
Package cucumber_demo_formatter implements a simple Cucumber formatter that prints
an emoji for each step.
*/

package cucumber_demo_formatter

import (
	"encoding/json"
	"fmt"
	"github.com/cucumber/common/messages/go/v17"
	"io"
)

func ProcessMessages(reader io.Reader, output io.Writer) {
	var emoji = map[messages.TestStepResultStatus]string{
		messages.TestStepResultStatus_UNKNOWN:   "👽",
		messages.TestStepResultStatus_PASSED:    "😃",
		messages.TestStepResultStatus_SKIPPED:   "🥶",
		messages.TestStepResultStatus_PENDING:   "⏰",
		messages.TestStepResultStatus_UNDEFINED: "🤷",
		messages.TestStepResultStatus_AMBIGUOUS: "🦄",
		messages.TestStepResultStatus_FAILED:    "💣",
	}

	decoder := json.NewDecoder(reader)
	for {
		envelope := &messages.Envelope{}
		err := decoder.Decode(envelope)
		if err == io.EOF {
			break
		}
		if err != nil {
			panic(err)
		}

		if envelope.TestStepFinished != nil {
			_, err = fmt.Fprintf(output, emoji[envelope.TestStepFinished.TestStepResult.Status])
		}
		if envelope.TestRunFinished != nil {
			_, err = fmt.Fprint(output, "\n")
		}
		if err != nil {
			panic(err)
		}
	}
}

File: demo-formatter/ruby/lib/cucumber_demo_formatter.rb
require 'cucumber/messages'

class CucumberDemoFormatter
  def process_messages(message_enumerator, output)
    emoji = {
      'UNKNOWN'   => '👽',
      'PASSED'    => '😃',
      'SKIPPED'   => '🥶',
      'PENDING'   => '⏰',
      'UNDEFINED' => '🤷',
      'AMBIGUOUS' => '🦄',
      'FAILED'    => '💣',
    }
    message_enumerator.each do |message|
      if message.test_step_finished
        status = message.test_step_finished.test_step_result.status
        em = emoji[status]
        raise "No emoji found for status #{status}" if em.nil?
        output.write(em)
      end
      if message.test_run_finished
        output.write("\n")
      end
    end
  end
end
